### PR TITLE
Compute routes backward

### PR DESCRIPTION
**Pull request by araspitzu**
Compute the routes backward from the target and with the correct channels cost, nodes that are 'neighbor' to the local node (direct channel) will have a routing cost of 0.

### CODE REVIEW

**Comment by araspitzu** [comment]
I chose to have an explicit parameter on the graph construction and manipulation functions, alternatively we can have a separate function with reverse in the name i.e "addEdgeReverse", please let me know what do you think of it.

**Comment by sstone** [comment]
Isn't it possible to use the same graph (i.e. not need for the additional `reverse` parameter) and just compute the route backwards internally ? It should be just an implementation choice that is not even exposed in the API ?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 1]
```diff
@@ -791,7 +791,7 @@ object Router {
   def findRoute(g: DirectedGraph, localNodeId: PublicKey, targetNodeId: PublicKey, amountMsat: Long, extraEdges: Set[GraphEdge] = Set.empty, ignoredEdges: Set[ChannelDesc] = Set.empty): Try[Seq[Hop]] = Try {
     if (localNodeId == targetNodeId) throw CannotRouteToSelf
 
-    Graph.shortestPath(g, localNodeId, targetNodeId, amountMsat, ignoredEdges, extraEdges) match {
+    Graph.shortestPath(g, targetNodeId, localNodeId, amountMsat, ignoredEdges.map(reverseDesc), extraEdges.map(edge => edge.copy(desc = reverseDesc(edge.desc)))) match {
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: I think we should keep the method and parameters as they were, we are still computing a shortest path from the source to the destination, it's just that internally we compute it backwards

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 2]
```diff
@@ -90,7 +90,9 @@ object Graph {
             val neighbor = edge.desc.b
 
             // note: the default value here will never be used, as there is always an entry for the current in the 'cost' map
-            val newMinimumKnownCost = cost.get(current.key) + edgeWeightByAmount(edge, amountMsat)
+            // note: when searching on a reversed graph we consider 0 the weight of the edge of the target-neighboring node
+            // this is because it will be a direct channel to the source and it pays no fees
+            val newMinimumKnownCost = cost.get(current.key) + edgeWeightByAmount(edge, amountMsat, neighbor == targetNode)
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: I don't think this is right: the fees that are applied are the fees for the outgoing edge, and which is then added to the amount that we want to forward. Suppose we have:

A
    \
     C -- D
  /
B
And want to forward X from A to D.
First we look for incoming edges of D that can carry X. We have one: edge_cd
Then we move backwards and look for incoming edges of C that can carry amount + cost(edge_cd, amount) and so on: the amount that is used to compute the relay fee is updated at each hop.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 3]
```diff
@@ -81,16 +81,17 @@ object Graph {
...
 
-            val neighbor = edge.desc.b
+          // note: 'cost' contains the smallest known cumulative cost (amount + fees) necessary to reach 'current' so far
+          // note: there is always an entry for the current in the 'cost' map
+          val newMinimumKnownCost = edgeWeight(edge, cost.get(current.key), neighbor == targetNode)
 
-            // note: the default value here will never be used, as there is always an entry for the current in the 'cost' map
-            val newMinimumKnownCost = cost.get(current.key) + edgeWeightByAmount(edge, amountMsat)
+          // test for ignored edges
+          if (!(edge.update.htlcMaximumMsat.exists(_ < newMinimumKnownCost) ||
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: is that the most optimal order for the testing conditions? Given that the first to return `true` will return immediately

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 4]
```diff
@@ -130,17 +131,25 @@ object Graph {
...
-  private def edgeWeightByAmount(edge: GraphEdge, amountMsat: Long): Long = {
-    nodeFee(edge.update.feeBaseMsat, edge.update.feeProportionalMillionths, amountMsat)
+  /**
+    *
+    * @param edge the edge for which we want to compute the weight
+    * @param amountWithFees the value that this edge will have to carry along
+    * @param isNeighborTarget true if the receiving vertex of this edge is the target node (source in a reversed graph), which has cost 0
+    * @return the new amount updated with the necessary fees for this edge
+    */
+  private def edgeWeight(edge: GraphEdge, amountWithFees: Long, isNeighborTarget: Boolean): Long = isNeighborTarget match {
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: would an `if` be more optimal here?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 5]
```diff
@@ -161,35 +170,42 @@ object Graph {
...
+        * Edges are added as "reversed" swapping the starting vertex with the final.
         *
         * @param edge the edge that is going to be added to the graph
         * @return a new graph containing this edge
         */
       def addEdge(edge: GraphEdge): DirectedGraph = {
 
-        val vertexIn = edge.desc.a
-        val vertexOut = edge.desc.b
+        val finalEdge = edge.copy(desc = reverseDesc(edge.desc))
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: how significant is this extra copy during graph initialization on mobile?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala [hunk 6]
```diff
@@ -308,18 +335,22 @@ object Graph {
...
+            case None => mutableMap += someDesc.b -> List.empty[GraphEdge]
             case _ =>
           }
         }
 
         new DirectedGraph(mutableMap.toMap)
       }
 
-      def graphEdgeToHop(graphEdge: GraphEdge): Hop = Hop(graphEdge.desc.a, graphEdge.desc.b, graphEdge.update)
+      def reverseDesc(desc: ChannelDesc): ChannelDesc = desc.copy(a = desc.b, b = desc.a)
```
Comment (pm47) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: Nitpick: I would define this as a member function of `ChannelDesc`

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala (hunk 4 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: Depends on which one you consider more readable, i've used 'match' for testing conditions and this is consistent to it.

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala (hunk 5 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: We don't call this method during the graph initialization, we use DirectedGraph.makeGraph...which also makes a copy of the edge desc

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala (hunk 3 above)
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala]: done

**Comment by araspitzu** [comment]
wiki updated

**Pull request merged**